### DISCUSSION TITLE
Fast Access/Indexing

### DISCUSSION THREAD

**Discussion by fullben** [JanWiemer/jacis#29] [General]
Access to objects in stores is really fast whenever one has the corresponding key. Now especially in stores with many objects, accessing objects based on some property or condition other than the key can be rather slow.
Consider the following example:

```java
public class User {
  private String id;
  private String username;
  ...
}

...
JacisStore<String, User> userStore = ...
// Access via id
User user1 = store.get(someId);
// Access via username (slower)
User user2 = store.stream(u -> u.getUsername().equals(someUsername)).findAny().orElse(null);
// What I would want to do in this situation
User user3 = store.getByUsername(someUsername);
```

I guess what I'm looking for are indexes... :)
Are there currently any library features which would allow me to emulate indexes somehow? If not, would it be possible to extend the existing store implementation to include custom indexes?

**Comment by JanWiemer** [comment]
Right you are looking for indexes, and you are not the first requesting for direct index support.
Probably I will consider to add some features in this direction to the JACIS store.
For now it should be possible to implement an index using the tracked view feature (see https://github.com/JanWiemer/jacis/wiki/Tracked-Views) with moderate effort.
The tracked view should maintain a map assigning the index key the name of the corresponding object (for unique indices) or the list of names of all corresponding objects (fo non unique indices). On each modification this map has to be maintained...

**Comment by JanWiemer** [comment]
Added support for (unique or non-unique) indices. See issue #32.
